muxink/src/codec/bytesrepr.rs

```diff
@@ -0,0 +1,172 @@
+where
+    T: ToBytes,
+{
+    type Error = Error;
+
+    type Output = Bytes;
+
+    fn transcode(&mut self, input: T) -> Result<Self::Output, Self::Error> {
+        Ok(input.to_bytes().map_err(Error::BytesreprError)?.into())
+    }
+}
+
+/// Bytesrepr decoder.
+#[derive(Default)]
+pub struct BytesreprDecoder<T> {
+    item_type: PhantomData<T>,
+}
+
+impl<T> BytesreprDecoder<T> {
+    /// Creates a new bytesrepr decoder.
+    pub fn new() -> Self {
+        BytesreprDecoder {
+            item_type: PhantomData,
+        }
+    }
+}
+
+impl<R, T> Transcoder<R> for BytesreprDecoder<T>
+where
+    T: FromBytes + Send + Sync + 'static,
+    R: AsRef<[u8]>,
+{
+    type Error = Error;
+
+    type Output = T;
+
+    fn transcode(&mut self, input: R) -> Result<Self::Output, Self::Error> {
+        Ok(bytesrepr::deserialize_from_slice(input).map_err(Error::BytesreprError)?)
+    }
+}
+
+impl<T> FrameDecoder for BytesreprDecoder<T>
+where
+    T: FromBytes + Send + Sync + 'static,
+{
+    type Error = Error;
+    type Output = T;
+
+    fn decode_frame(&mut self, buffer: &mut BytesMut) -> DecodeResult<Self::Output, Self::Error> {
+        let transcoded = FromBytes::from_bytes(buffer.as_ref());
+        match transcoded {
+            Ok((data, rem)) => {
+                let _ = buffer.split_to(buffer.remaining() - rem.len());
+                DecodeResult::Item(data)
+            }
+            Err(err) => match &err {
+                bytesrepr::Error::EarlyEndOfStream => DecodeResult::Incomplete,
+                bytesrepr::Error::Formatting
+                | bytesrepr::Error::LeftOverBytes
+                | bytesrepr::Error::NotRepresentable
+                | bytesrepr::Error::ExceededRecursionDepth
+                | bytesrepr::Error::OutOfMemory => Failed(Error::BytesreprError(err)),
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{DecodeResult, Error};
+    use crate::codec::{
+        bytesrepr::{BytesreprDecoder, BytesreprEncoder},
+        BytesMut, FrameDecoder, Transcoder,
+    };
+    use casper_types::bytesrepr;
+
+    #[test]
+    fn roundtrip() {
+        let data = "abc";
+
+        let mut encoder = BytesreprEncoder::new();
+        let value: String = String::from(data);
+        let encoded = encoder.transcode(value).expect("should encode");
+
+        let mut decoder = BytesreprDecoder::<String>::new();
+        let decoded = decoder.transcode(encoded).expect("should decode");
+
+        assert_eq!(data, decoded);
+    }
+
+    #[test]
+    fn decodes_frame() {
+        let data = b"\x03\0\0\0abc\x04\0\0\0defg";
+
+        let mut bytes: BytesMut = BytesMut::new();
+        bytes.extend(data);
+
+        let mut decoder = BytesreprDecoder::<String>::new();
+
+        assert!(matches!(decoder.decode_frame(&mut bytes), DecodeResult::Item(i) if i == "abc"));
+        assert!(matches!(decoder.decode_frame(&mut bytes), DecodeResult::Item(i) if i == "defg"));
+    }
+
```

Comment:
See comments on `bincode` tests about generalizing this.

Also, this also should have the mentioned test added that is a "naked stream" of integers.

Reply:
Regarding generalization, see [this comment](https://github.com/casper-network/casper-node/pull/3220#discussion_r932975507).

And the mentioned test is added in 04b601a8b98e343e30591a0a0ce258d583550c84